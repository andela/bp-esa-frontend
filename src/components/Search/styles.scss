.search-container{
  display: flex;
  flex-direction: row;
  margin-top: 15px;
  height: 40px;
}

.search-input{
    margin-left: 1.7%;
    padding: 8px 20px;
    width: 50%;
    border-radius: 3px;
    border: 1px solid grey;
    text-transform:capitalize;
    transition: border 0.5s;
    &:focus{
      outline: none;
      border: 2px solid lighten(grey, 20%);
    }
    &:hover{
      border: 2px solid lighten(grey, 20%);
    }
}

.search {
    font-size: 14px;
    margin-left: 2%;

    .search-title {
      display: flex;
      flex-direction: row;
      font-family: 'Din Pro';
      height: 100%;
      padding: 8.5px 20px;
      border: 1px solid grey;
      border-radius: 3px;
      color: #4d4d4d;
      cursor: pointer;
      transition-property: color;
      &:hover {
        background: lighten(#385de0, 45%);
      }
      i {
        padding-top: 7.4px;
        margin: -6px 0px -6px 10px;
        font-size: 20px;
        transition: all 0.3s;
        &.rotate-180 {
          transform: rotate(180deg);
      }
    }
  }
      .search-option{
        visibility: hidden;
        margin-top: 5px;
        margin-left: 3px;
        box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
        width: 97%;
        border-radius: 6px;
        overflow: hidden;
        li {
          z-index:1;
          font-family: 'Din Pro';
          background-color:  white;
          border-bottom: 1px solid #d9d9d9;
          color: #4d4d4d;
          padding: 5px;
          input[type=radio] {
            margin-right: 10px;
          }
        }
        &.search-options-isvisible {
          visibility: visible;
        }
      }
}
