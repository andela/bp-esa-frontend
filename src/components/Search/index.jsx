
import React, { PureComponent } from 'react';
import PropTypes from 'proptypes';
import './styles.scss';
import classNames from 'classnames';
import ReactDOM from 'react-dom';

class Search extends PureComponent {
  constructor() {
    super();
    this.state = {
      searchOptionsVisible: false,
      searchValue: '',
      searchCriteria: 1,
    };
  }

  componentDidUpdate() {
    const { searchValue, searchCriteria } = this.state;
    const { handleSearch } = this.props;
    if (searchValue || searchCriteria) {
      handleSearch(searchValue, searchCriteria);
    }
  }


  toggleVisibility = (e) => {
    // The DOM's UL element holding the list of options
    const searchOptionsEl = ReactDOM.findDOMNode(this)
      .getElementsByClassName('search-option')[0];
    const clickOff = (ev) => {
      if (ev.target === searchOptionsEl) return;
      if (e !== ev) {
        this.setState(() => ({ searchOptionsVisible: false }));
        document.removeEventListener('click', clickOff);
      }
    };

    if (!this.state.searchOptionsVisible) {
      this.setState(() => ({ searchOptionsVisible: true }));
      document.addEventListener('click', clickOff);
    }
  }

  handleSearchCriteriaChange = (optionId) => {
    this.setState({
      searchCriteria: optionId,
    });
  };

  handleSearchValueChange = (event) => {
    event.preventDefault();
    this.setState({ searchValue: event.target.value });
  }

  renderSearchOptions() {
    const { searchOptionsVisible } = this.state;
    return (
      <ul className={classNames('search-option', { 'search-options-isvisible': searchOptionsVisible })}>
        <li>
          <input type="radio" defaultChecked name="search" onChange={() => this.handleSearchCriteriaChange(1)} />
          {'Fellow Name'}
        </li>
        <li>
          <input type="radio" name="search" onChange={() => this.handleSearchCriteriaChange(2)} />
          {'Partner Name'}
        </li>
      </ul>
    );
  }

  render() {
    const { searchValue } = this.state;
    const { searchOptionsVisible, searchCriteria } = this.state;
    return (
      <div className="search-container">
        <input type="text" className="search-input" value={searchValue} onChange={this.handleSearchValueChange} />
        <div className="search">
          <div className="search-title" onClick={event => this.toggleVisibility(event)}>
            <span className="title">
              {'Search With'}
              {' | '}
              {searchCriteria === 1 ? 'Fellow Name' : 'Partner Name'}
            </span>

            <i
              className={classNames('fa fa-angle-down', { 'rotate-180': searchOptionsVisible })}
            />
          </div>
          {this.renderSearchOptions()}
        </div>
      </div>
    );
  }
}

Search.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};

export default Search;
