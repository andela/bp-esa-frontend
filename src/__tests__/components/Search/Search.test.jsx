import React from 'react';
import Search from '../../../components/Search';

const props = {
  handleSearch: jest.fn(),
};

const getComponent = () => shallow(<Search {...props} />);


describe('<Search />', () => {
  it('should render as expected', () => {
    expect(getComponent()).toMatchSnapshot();
  });

  describe('toggleVisibility method', () => {
    it('should initially set the searchOptionsVisible to "FALSE"', () => {
      const component = (() => shallow(<Search {...props} />))();
      expect(component.state('searchOptionsVisible')).toBeFalsy();
    })

    it("should change the searchOptionsVisible in the state", () => {
      const mountedComponent = mount(<Search {...props} />);
      const mountedComponentInstance = mountedComponent.instance();
      const isVisible = mountedComponent.state('searchOptionsVisible');
      mountedComponentInstance.toggleVisibility();
      expect(isVisible).not.toEqual(mountedComponent.state("searchOptionsVisible"));
    });
    
    it('should change the searchOptionsVisible in the state when dropdown is clicked', () => {
      const component = (() => mount(<Search {...props} />))();
      component.find('.search-title').simulate('click');
      expect(component.state('searchOptionsVisible')).toBeTruthy();
    });

    it('should have a single instance of optionSet in Search dropdown', () => {
      const component = mount(<Search {...props} />);
      component.find('.search-title').simulate('click');
      expect(component.find('.search-options-isvisible').length).toBe(1);
    });
    
    it('should close dropdown options when user clicks away', () => {
      const newprops = {...props, options: [{ type: null }]};
      const component = mount(<Search {...newprops} />);
      component.find('.search-title').simulate('click');
      document.body.click();
      expect(component.state('searchOptionsVisible')).toBeFalsy();
    });

    it('should open dropdown when called', () => {
      const newprops = {...props, options: [{ type: null }]};
      const component = mount(<Search {...newprops} />);
      const componentInstance = component.instance();
      componentInstance.toggleVisibility();
      expect(component.state('searchOptionsVisible')).toBeTruthy();
    });

  });

  describe('handleSearch method', () => {
    it(`should render the search options
    `, () => {
      const component = getComponent();
      expect(component.state('searchValue')).toEqual('');
      expect(component.state('searchCriteria')).toEqual(1);
      component.setState({
        searchCriteria: 1,
      });
      component.setState({
        searchValue: 'Andela',
      });
      expect(props.handleSearch).toHaveBeenCalled();
    });

    it('should call handleSearchCriteriaChange', () => {
      const component = getComponent();
      component.find('input[name="search"]').last().simulate('change');
      expect(component.state().searchCriteria).toBe(2);
      component.find('input[name="search"]').first().simulate('change');
      expect(component.state().searchCriteria).toBe(1);
    });

    it('should call handleSearchValueChange', () => {
      const component = getComponent();
      const event = {
        preventDefault: jest.fn(),
        target: {
          value: 'Somevalue',
        },
      };
      component.find('.search-input').simulate('change', event);
      expect(component.state().searchValue).toBe('Somevalue');
    });
  });
});
