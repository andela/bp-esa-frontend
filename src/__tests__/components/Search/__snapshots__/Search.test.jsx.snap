// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    handleSearch={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="search-input"
          onChange={[Function]}
          type="text"
          value=""
        />,
        <div
          className="search"
        >
          <div
            className="search-title"
            onClick={[Function]}
          >
            <span
              className="title"
            >
              Search With
               | 
              Fellow Name
            </span>
            <i
              className="fa fa-angle-down"
            />
          </div>
          <ul
            className="search-option"
          >
            <li>
              <input
                defaultChecked={true}
                name="search"
                onChange={[Function]}
                type="radio"
              />
              Fellow Name
            </li>
            <li>
              <input
                name="search"
                onChange={[Function]}
                type="radio"
              />
              Partner Name
            </li>
          </ul>
        </div>,
      ],
      "className": "search-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "search-input",
          "onChange": [Function],
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="search-title"
              onClick={[Function]}
            >
              <span
                className="title"
              >
                Search With
                 | 
                Fellow Name
              </span>
              <i
                className="fa fa-angle-down"
              />
            </div>,
            <ul
              className="search-option"
            >
              <li>
                <input
                  defaultChecked={true}
                  name="search"
                  onChange={[Function]}
                  type="radio"
                />
                Fellow Name
              </li>
              <li>
                <input
                  name="search"
                  onChange={[Function]}
                  type="radio"
                />
                Partner Name
              </li>
            </ul>,
          ],
          "className": "search",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="title"
                >
                  Search With
                   | 
                  Fellow Name
                </span>,
                <i
                  className="fa fa-angle-down"
                />,
              ],
              "className": "search-title",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Search With",
                    " | ",
                    "Fellow Name",
                  ],
                  "className": "title",
                },
                "ref": null,
                "rendered": Array [
                  "Search With",
                  " | ",
                  "Fellow Name",
                ],
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-angle-down",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <li>
                  <input
                    defaultChecked={true}
                    name="search"
                    onChange={[Function]}
                    type="radio"
                  />
                  Fellow Name
                </li>,
                <li>
                  <input
                    name="search"
                    onChange={[Function]}
                    type="radio"
                  />
                  Partner Name
                </li>,
              ],
              "className": "search-option",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <input
                      defaultChecked={true}
                      name="search"
                      onChange={[Function]}
                      type="radio"
                    />,
                    "Fellow Name",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "defaultChecked": true,
                      "name": "search",
                      "onChange": [Function],
                      "type": "radio",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "Fellow Name",
                ],
                "type": "li",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <input
                      name="search"
                      onChange={[Function]}
                      type="radio"
                    />,
                    "Partner Name",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "name": "search",
                      "onChange": [Function],
                      "type": "radio",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "Partner Name",
                ],
                "type": "li",
              },
            ],
            "type": "ul",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="search-input"
            onChange={[Function]}
            type="text"
            value=""
          />,
          <div
            className="search"
          >
            <div
              className="search-title"
              onClick={[Function]}
            >
              <span
                className="title"
              >
                Search With
                 | 
                Fellow Name
              </span>
              <i
                className="fa fa-angle-down"
              />
            </div>
            <ul
              className="search-option"
            >
              <li>
                <input
                  defaultChecked={true}
                  name="search"
                  onChange={[Function]}
                  type="radio"
                />
                Fellow Name
              </li>
              <li>
                <input
                  name="search"
                  onChange={[Function]}
                  type="radio"
                />
                Partner Name
              </li>
            </ul>
          </div>,
        ],
        "className": "search-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "search-input",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="search-title"
                onClick={[Function]}
              >
                <span
                  className="title"
                >
                  Search With
                   | 
                  Fellow Name
                </span>
                <i
                  className="fa fa-angle-down"
                />
              </div>,
              <ul
                className="search-option"
              >
                <li>
                  <input
                    defaultChecked={true}
                    name="search"
                    onChange={[Function]}
                    type="radio"
                  />
                  Fellow Name
                </li>
                <li>
                  <input
                    name="search"
                    onChange={[Function]}
                    type="radio"
                  />
                  Partner Name
                </li>
              </ul>,
            ],
            "className": "search",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="title"
                  >
                    Search With
                     | 
                    Fellow Name
                  </span>,
                  <i
                    className="fa fa-angle-down"
                  />,
                ],
                "className": "search-title",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "Search With",
                      " | ",
                      "Fellow Name",
                    ],
                    "className": "title",
                  },
                  "ref": null,
                  "rendered": Array [
                    "Search With",
                    " | ",
                    "Fellow Name",
                  ],
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "fa fa-angle-down",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "i",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <li>
                    <input
                      defaultChecked={true}
                      name="search"
                      onChange={[Function]}
                      type="radio"
                    />
                    Fellow Name
                  </li>,
                  <li>
                    <input
                      name="search"
                      onChange={[Function]}
                      type="radio"
                    />
                    Partner Name
                  </li>,
                ],
                "className": "search-option",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <input
                        defaultChecked={true}
                        name="search"
                        onChange={[Function]}
                        type="radio"
                      />,
                      "Fellow Name",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "defaultChecked": true,
                        "name": "search",
                        "onChange": [Function],
                        "type": "radio",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "Fellow Name",
                  ],
                  "type": "li",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <input
                        name="search"
                        onChange={[Function]}
                        type="radio"
                      />,
                      "Partner Name",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "name": "search",
                        "onChange": [Function],
                        "type": "radio",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "Partner Name",
                  ],
                  "type": "li",
                },
              ],
              "type": "ul",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
